kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h

```diff
@@ -0,0 +1,793 @@
+     *  @param[in] rCurrentProcessInfo ProcessInfo instance for the containing ModelPart.
+     */
+    void AddDynamicsToRHS(Condition::Pointer rCurrentCondition,
+                          LocalSystemVectorType& rRHS_Contribution,
+                          LocalSystemMatrixType& rDampingMatrix,
+                          LocalSystemMatrixType& rMassMatrix,
+                          ProcessInfo& rCurrentProcessInfo)
+    {
+        // adding inertia contribution
+        if (rMassMatrix.size1() != 0 && mUpdateAcceleration)
+        {
+            int k = OpenMPUtils::ThisThread();
+            rCurrentCondition->GetSecondDerivativesVector(
+                mSecondDerivativeValuesVector[k], 0);
+            (mSecondDerivativeValuesVector[k]) *= (1.00 - mBossak.Alpha);
+            rCurrentCondition->GetSecondDerivativesVector(
+                mSecondDerivativeValuesVectorOld[k], 1);
+            noalias(mSecondDerivativeValuesVector[k]) +=
+                mBossak.Alpha * mSecondDerivativeValuesVectorOld[k];
+
+            noalias(rRHS_Contribution) -=
+                prod(rMassMatrix, mSecondDerivativeValuesVector[k]);
+        }
+    }
+
+    void UpdateTimeSchemeVariables(ModelPart& rModelPart)
+    {
+        KRATOS_TRY;
+
+        UpdateAcceleration(rModelPart);
+        UpdateDisplacement(rModelPart);
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+private:
+    struct BossakConstants
+    {
+        double Alpha;
+        double Gamma;
+        double Beta;
+        double C0;
+        double C1;
+        double C2;
+        double C3;
+        double C4;
+        double C5;
+        double C6;
+    };
+
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    BossakConstants mBossak;
+
+    typename TSparseSpace::DofUpdaterPointerType mpDofUpdater =
+        TSparseSpace::CreateDofUpdater();
+
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectCurrentVelocityVector;
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectCurrentAccelerationVector;
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectCurrentDisplacementVector;
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectOldVelocityVector;
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectOldAccelerationVector;
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectOldDisplacementVector;
+    std::vector<std::vector<Dof<double>::Pointer>> mVelocityDofVector;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    void InitializeNodeNeighbourCount(ModelPart::NodesContainerType& rNodes)
+    {
+        // This loop should not be omp parallel
+        // The operation is not threadsafe if the value is uninitialized
```

Comment:
nope, SetValue is threadsafe
GetValue is not

check my old issues, I discussed this with the others already

btw you should in any case use the variableutils for this